File: src/jira_dc_mcp/tools/boards.py
"""Agile board introspection tools."""

from __future__ import annotations

import json

from ..client import JiraClient


async def list_boards(client: JiraClient, project_key: str | None = None) -> str:
    """List all agile boards, optionally filtered by project."""
    boards = await client.list_boards(project_key=project_key)
    result = [
        {
            "id": b.get("id"),
            "name": b.get("name"),
            "type": b.get("type"),
            "projectKey": (b.get("location") or {}).get("projectKey"),
            "projectName": (b.get("location") or {}).get("projectName"),
        }
        for b in boards
    ]
    return json.dumps(result, indent=2)


async def get_board_configuration(client: JiraClient, board_id: int) -> str:
    """Get board configuration: columns, estimation, ranking, filter."""
    config = await client.get_board_configuration(board_id)

    columns = []
    for col in (config.get("columnConfig") or {}).get("columns", []):
        columns.append({
            "name": col.get("name"),
            "statuses": [s.get("id") for s in col.get("statuses", [])],
            "min": col.get("min"),
            "max": col.get("max"),
        })

    estimation = config.get("estimation") or {}
    ranking = config.get("ranking") or {}
    filter_ref = config.get("filter") or {}

    result = {
        "id": config.get("id"),
        "name": config.get("name"),
        "type": config.get("type"),
        "filter": {
            "id": filter_ref.get("id"),
            "name": filter_ref.get("name"),
            "query": filter_ref.get("query"),
        },
        "columnConfig": {
            "constraintType": (config.get("columnConfig") or {}).get("constraintType"),
            "columns": columns,
        },
        "estimation": {
            "type": estimation.get("type"),
            "field": estimation.get("field", {}).get("displayName") if estimation.get("field") else None,
        },
        "ranking": {
            "rankCustomFieldId": ranking.get("rankCustomFieldId"),
        },
    }

    # Include sub-query (swimlane/quick filter) if present
    sub_query = config.get("subQuery")
    if sub_query:
        result["subQuery"] = sub_query

    return json.dumps(result, indent=2)

File: src/tools/boards.ts
/** Agile board introspection tools. */

import { z } from "zod";
import { dumps } from "../json.js";
import type { ToolDef } from "./types.js";

export const boardTools: ToolDef[] = [
  {
    name: "list_boards",
    description:
      "List all agile boards (Scrum/Kanban). Optionally filter by project key. " +
      "Shows board name, type, and associated project.",
    inputShape: {
      project_key: z
        .string()
        .optional()
        .describe("Optional project key to filter boards. Omit for all boards."),
    },
    async handler({ client }, args) {
      const boards = await client.listBoards(args.project_key);
      return dumps(
        boards.map((b: any) => ({
          id: b.id,
          name: b.name,
          type: b.type,
          projectKey: (b.location ?? {}).projectKey,
          projectName: (b.location ?? {}).projectName,
        })),
      );
    },
  },

  {
    name: "get_board_configuration",
    description:
      "Get board configuration: columns (with status mappings and WIP limits), " +
      "estimation settings, ranking, and backing filter/JQL. " +
      "Shows how teams actually see and manage work vs. the underlying workflow.",
    inputShape: { board_id: z.coerce.number().int().describe("Agile board ID") },
    async handler({ client }, args) {
      const config = await client.getBoardConfiguration(args.board_id);

      const columns = ((config.columnConfig ?? {}).columns ?? []).map((col: any) => ({
        name: col.name,
        statuses: (col.statuses ?? []).map((s: any) => s.id),
        min: col.min,
        max: col.max,
      }));

      const estimation = config.estimation ?? {};
      const ranking = config.ranking ?? {};
      const filterRef = config.filter ?? {};

      const result: Record<string, unknown> = {
        id: config.id,
        name: config.name,
        type: config.type,
        filter: {
          id: filterRef.id,
          name: filterRef.name,
          query: filterRef.query,
        },
        columnConfig: {
          constraintType: (config.columnConfig ?? {}).constraintType,
          columns,
        },
        estimation: {
          type: estimation.type,
          field: estimation.field ? (estimation.field.displayName ?? null) : null,
        },
        ranking: {
          rankCustomFieldId: ranking.rankCustomFieldId,
        },
      };

      if (config.subQuery && Object.keys(config.subQuery).length > 0) {
        result.subQuery = config.subQuery;
      }
      return dumps(result);
    },
  },
];
